p2p/discovery/discovery.go

```diff
@@ -0,0 +1,116 @@
+// Package discovery implements a Distributed Hash Table based on Kademlia protocol.
+package discovery
+
+import (
+	"context"
+	"errors"
+	"github.com/spacemeshos/go-spacemesh/p2p/config"
+	"github.com/spacemeshos/go-spacemesh/p2p/node"
+	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+	"github.com/spacemeshos/go-spacemesh/p2p/server"
+)
+
+// Interface is an interface to the discovery protocol
+type Interface interface {
+	Remove(pubkey p2pcrypto.PublicKey)
+	Lookup(pubkey p2pcrypto.PublicKey) (node.Node, error)
+	Update(addr, src node.Node)
+	SelectPeers(qty int) []node.Node
+	Bootstrap(ctx context.Context) error
+	Size() int
+	SetLocalAddresses(tcp, udp string)
+}
+
+type Protocol interface {
+	Ping(p p2pcrypto.PublicKey) error
+	GetAddresses(server p2pcrypto.PublicKey) ([]discNode, error)
+	SetLocalAddresses(tcp, udp string)
+}
+
+var (
+	// ErrLookupFailed determines that we could'nt find this node in the routing table or network
+	ErrLookupFailed = errors.New("failed to find node in the network")
+	// ErrEmptyRoutingTable means that our routing table is empty thus we can't find any node (so we can't query any)
+	ErrEmptyRoutingTable = errors.New("no nodes to query - routing table is empty")
+)
+
+// Discovery is struct that holds the protocol components, the protocol definition, the addr book data structure and more.
+type Discovery struct {
+	config config.SwarmConfig
+
+	disc Protocol
+
+	local        *node.LocalNode
+	rt           *addrBook
+	bootstrapper *refresher
+}
+
+// Size returns the size of addrBook.
+func (d *Discovery) Size() int {
+	return d.rt.NumAddresses()
+}
+
+// SelectPeers asks routing table to randomly select a slice of nodes in size `qty`
+func (d *Discovery) SelectPeers(qty int) []node.Node {
```

Comment:
what happens if no peers are found? is this expected and normal behaviour?

Reply:
yes this is handled because we might actually have an empty table